File: templates/python/xor_algorithm.py
"""
XOR Collaborative Editor — Python Algorithm Template
Real-time collaborative mathematical utilities & bitwise cryptographic hashing.
"""

from typing import List, Tuple, Dict, Any
import hashlib
import time


class CollaborativeXORSession:
    """
    Simulates collaborative operational transform & bitwise XOR algorithms
    for synchronized pair-programming rooms.
    """

    def __init__(self, room_id: str, host_name: str) -> None:
        self.room_id: str = room_id
        self.host_name: str = host_name
        self.collaborators: List[str] = [host_name]
        self.history: List[Dict[str, Any]] = []

    def compute_xor_checksum(self, data: bytes) -> int:
        """Computes rolling bitwise XOR checksum of a byte buffer."""
        checksum = 0
        for byte in data:
            checksum ^= byte
        return checksum

    def bitwise_xor_cipher(self, message: str, key: str) -> str:
        """
        Symmetric XOR string obfuscation for private room payload transmission.
        """
        cipher_chars = []
        for i, char in enumerate(message):
            key_char = key[i % len(key)]
            cipher_chars.append(chr(ord(char) ^ ord(key_char)))
        return "".join(cipher_chars)

    def add_peer(self, peer_name: str) -> bool:
        """Registers a new peer in the collaborative editing room."""
        if peer_name not in self.collaborators:
            self.collaborators.append(peer_name)
            self.history.append({
                "action": "PEER_JOIN",
                "user": peer_name,
                "timestamp": time.time()
            })
            return True
        return False

    def get_room_summary(self) -> Dict[str, Any]:
        """Returns structured JSON summary of active room telemetry."""
        return {
            "roomId": self.room_id,
            "host": self.host_name,
            "activeUsersCount": len(self.collaborators),
            "users": self.collaborators,
            "totalRevisions": len(self.history)
        }


def find_unique_element(nums: List[int]) -> int:
    """
    Classic XOR property demonstration:
    x ^ x = 0, x ^ 0 = x
    Finds the single non-duplicate number in a list of pairs in O(N) time and O(1) space.
    """
    unique = 0
    for num in nums:
        unique ^= num
    return unique


if __name__ == "__main__":
    session = CollaborativeXORSession("xor-alpha-99", "Nidhi")
    session.add_peer("Alice")
    session.add_peer("Bob")

    test_data = [4, 1, 2, 1, 2, 4, 99]
    print(f"[*] XOR Single Element in {test_data}: {find_unique_element(test_data)}")

    secret = "XOR Live Collaborative Code 2026"
    key = "pink_theme_secret"
    encrypted = session.bitwise_xor_cipher(secret, key)
    decrypted = session.bitwise_xor_cipher(encrypted, key)

    print(f"[*] Original:  {secret}")
    print(f"[*] Decrypted: {decrypted}")
    print(f"[*] Room Status: {session.get_room_summary()}")

File: templates/cpp/xor_algorithm.cpp
/**
 * XOR Collaborative Editor — C++ Algorithm & Data Structure Template
 * Multi-threaded room buffer synchronization & Bitwise bit manipulation routines.
 */

#include <iostream>
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>
#include <iomanip>

class CollaborativeRoomManager {
private:
    std::string roomId;
    std::vector<std::string> activePeers;
    std::vector<int> documentBuffer;

public:
    CollaborativeRoomManager(const std::string& id, const std::string& host) 
        : roomId(id) {
        activePeers.push_back(host);
    }

    void addPeer(const std::string& peer) {
        if (std::find(activePeers.begin(), activePeers.end(), peer) == activePeers.end()) {
            activePeers.push_back(peer);
        }
    }

    // Classic XOR single-number algorithm: O(N) time, O(1) space
    int findSingleNumber(const std::vector<int>& numbers) const {
        int result = 0;
        for (int num : numbers) {
            result ^= num;
        }
        return result;
    }

    // Fast XOR Swap demonstration
    void fastXorSwap(int& a, int& b) const {
        if (&a == &b) return;
        a ^= b;
        b ^= a;
        a ^= b;
    }

    // Byte XOR parity check
    uint8_t computeParity(const std::string& text) const {
        uint8_t parity = 0;
        for (char c : text) {
            parity ^= static_cast<uint8_t>(c);
        }
        return parity;
    }

    void displayStatus() const {
        std::cout << "========================================\n";
        std::cout << "  🌸 XOR C++ Engine Session: " << roomId << "\n";
        std::cout << "  👥 Connected Peers (" << activePeers.size() << "):\n";
        for (const auto& peer : activePeers) {
            std::cout << "     - " << peer << "\n";
        }
        std::cout << "========================================\n";
    }
};

int main() {
    CollaborativeRoomManager room("xor-session-live", "Nidhi");
    room.addPeer("Alice");
    room.addPeer("Developer2");

    room.displayStatus();

    std::vector<int> dataset = {7, 3, 5, 4, 5, 3, 4};
    int unique = room.findSingleNumber(dataset);
    std::cout << "[*] Single non-paired element: " << unique << "\n";

    int x = 42, y = 137;
    std::cout << "[*] Before XOR Swap: x=" << x << ", y=" << y << "\n";
    room.fastXorSwap(x, y);
    std::cout << "[*] After XOR Swap:  x=" << x << ", y=" << y << "\n";

    std::string sample = "XOR Collaborative Live";
    std::cout << "[*] Parity Checksum: 0x" 
              << std::hex << std::uppercase 
              << static_cast<int>(room.computeParity(sample)) << "\n";

    return 0;
}
